### PR TITLE
3529-2

### PR DESCRIPTION
**Pull request by gannagoodkevich**
## Номер
3529

## Номер задания
2

## Ссылка на видео с демо
https://youtu.be/FSavG_8Bc1g

### CODE REVIEW

File: 3529/2/Data/camp_participants.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+---
+participents:
+- 3529: 
+  telegram_id: 313528695
```
Comment (DmitryAvramec) [inline 3529/2/Data/camp_participants.yaml]: O_o

File: 3529/2/checkin_context.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+module CheckinContext
+  include ImageFileWriter
+  include CoordinatesHandler
+  include SelfieHandler
+
+  def checkin!(message = nil, *)
+    save_context :checkin!
+    @path_check = PathLoader.new(payload)
+    new = message
```
Comment (DmitryAvramec) [inline 3529/2/checkin_context.rb]: new?

File: 3529/2/checkin_context.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    if new
```
Comment (DmitryAvramec) [inline 3529/2/checkin_context.rb]: Rewrite it without `if ... else`

File: 3529/2/checkin_context.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    else
+      reply
+    end
+  end
+
+  private
+
+  def reply
+    if payload['text']
+      respond_with :message, text: 'Send me your selfie, please'
```
Comment (DmitryAvramec) [inline 3529/2/checkin_context.rb]: wtf?

File: 3529/2/checkout_context.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    if new
+      coord_handle(@path_check, 'out')
+    else
+      replying
+    end
+  end
+
+  private
+
+  def replying
```
Comment (DmitryAvramec) [inline 3529/2/checkout_context.rb]: And why it's not `reply`? Same as above

File: 3529/2/coordinates_handler.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+require_relative 'text_file_writer'
+
+module CoordinatesHandler
+  include TextFileWriter
+
+  def coord_handle(path, in_or_out)
+    path_txt = path.create_directory(in_or_out, "#{payload['from']['username']}.txt")
+    write_text_file(path_txt)
```
Comment (DmitryAvramec) [inline 3529/2/coordinates_handler.rb]: Why it's txt?

File: 3529/2/image_loader.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+module ImageLoader
+  def load_image
+    file_id =  payload['photo'].first['file_id']
+    response = HTTParty.get("https://api.telegram.org/bot#{TOKEN}/getFile?file_id=#{file_id}")
+    "https://api.telegram.org/file/bot#{TOKEN}/#{response['result']['file_path']}"
```
Comment (DmitryAvramec) [inline 3529/2/image_loader.rb]: Constants?

File: 3529/2/number_checker.rb [hunk 8]
```diff
@@ -0,0 +1,44 @@
+require 'yaml'
+require 'redis'
+require 'redis-activesupport'
+require 'redis-rails'
```
Comment (DmitryAvramec) [inline 3529/2/number_checker.rb]: You have rails here? O_o

File: 3529/2/number_checker.rb [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+require 'redis-namespace'
```
Comment (DmitryAvramec) [inline 3529/2/number_checker.rb]: Why this?

File: 3529/2/number_checker.rb [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+require 'redis-rack-cache'
+
+class NumberChecker
+  def initialize(file_path)
+    @file_path = file_path
+    @file = YAML.safe_load(File.read(@file_path))
+    @resp = ''
```
Comment (DmitryAvramec) [inline 3529/2/number_checker.rb]: ???

File: 3529/2/number_checker.rb [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+  end
+
+  def handle_number(number, payload)
+    @resp = ''
```
Comment (DmitryAvramec) [inline 3529/2/number_checker.rb]: ????? 2

File: 3529/2/number_checker.rb [hunk 12]
```diff
@@ -0,0 +1,44 @@
+require 'yaml'
+require 'redis'
+require 'redis-activesupport'
+require 'redis-rails'
+require 'redis-namespace'
+require 'redis-rack-cache'
+
+class NumberChecker
+  def initialize(file_path)
+    @file_path = file_path
```
Comment (DmitryAvramec) [inline 3529/2/number_checker.rb]: If you need `file_path` only for load YML you don't need this instance var

File: 3529/2/number_checker.rb [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    @file = YAML.safe_load(File.read(@file_path))
+    @resp = ''
+  end
+
+  def handle_number(number, payload)
+    @resp = ''
+    @file['participents'].each do |participant|
+      @resp = checker_id(participant, payload, number) if number == participant.keys.first.to_s
```
Comment (DmitryAvramec) [inline 3529/2/number_checker.rb]: Clarify this one

File: 3529/2/number_checker.rb [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+              "I'm sorry, but someone already have signed in with #{number} number"
+            else
+              new_part
+            end
+    participant['telegram_id'] = payload['from']['id'] if @resp.eql?(new_part)
+    @resp
+  end
+
+  def update_file
+    data_base = Redis::Namespace.new('telegram-bot-app', redis: Redis.new)
```
Comment (DmitryAvramec) [inline 3529/2/number_checker.rb]: Why the hell it's `Redis.new`!

File: 3529/2/path_loader.rb [hunk 15]
```diff
@@ -0,0 +1,15 @@
+require 'fileutils'
+
+class PathLoader
+  def initialize(payload)
+    @session_id = "#{payload['from']['id']}:#{payload['chat']['id']}"
```
Comment (DmitryAvramec) [inline 3529/2/path_loader.rb]: Why you need all payload here if you use only 2 args?

File: 3529/2/selfie_handler.rb [hunk 16]
```diff
@@ -0,0 +1,11 @@
+require_relative 'image_file_writer'
+
+module SelfieHandler
+  include ImageFileWriter
+
+  def selfie_handler(path, in_or_out)
+    path_img = path.create_directory(in_or_out, "#{payload['from']['username']}.jpg")
```
Comment (DmitryAvramec) [inline 3529/2/selfie_handler.rb]: And why you need to create a directory here? and as I see `path` is not a path to file or etc it's object of `PathLoader`

File: 3529/2/start_context.rb [hunk 17]
```diff
@@ -0,0 +1,27 @@
+require 'telegram/bot'
+require 'yaml'
+require_relative 'number_checker'
+
+module StartContext
+  def start!(message = nil, *)
+    save_context :start!
+    if message
+      answer = response(message)
+      respond_with :message, text: answer.to_s
```
Comment (DmitryAvramec) [inline 3529/2/start_context.rb]: `answer.to_s` ? why your method can't return string?

File: 3529/2/start_context.rb [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+    else
+      respond_with :message, text: 'Hello, rook! Enter your camp number'
+    end
+  end
+
+  private
+
+  def response(message)
+    number_checker = NumberChecker.new('Data/camp_participants.yaml')
```
Comment (DmitryAvramec) [inline 3529/2/start_context.rb]: constants?

File: 3529/2/telegram.rb [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+  include CheckoutContext
+  include Telegram::Bot::UpdatesController::MessageContext
+
+  def initialize(*)
+    super
+    Telegram::Bot::UpdatesController.session_store = :redis_store, { expires_in: 100_000_000 }
+  end
+end
+
+TOKEN = '919190207:AAFfJhW2frNEWYeaSAvxhgwC6I233JVnVBg'.freeze
```
Comment (DmitryAvramec) [inline 3529/2/telegram.rb]: ENV?

File: 3529/2/telegram.rb [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+require_relative 'checkin_context'
+require_relative 'checkout_context'
+
+class WebhooksController < Telegram::Bot::UpdatesController
+  include StartContext
+  include CheckinContext
+  include CheckoutContext
+  include Telegram::Bot::UpdatesController::MessageContext
+
+  def initialize(*)
```
Comment (DmitryAvramec) [inline 3529/2/telegram.rb]: why *

File: 3529/2/telegram.rb [hunk 21]
```diff
@@ -0,0 +1,30 @@
+require 'telegram/bot'
+require 'yaml'
+require 'httparty'
+require 'open-uri'
+require_relative 'start_context'
+require_relative 'checkin_context'
+require_relative 'checkout_context'
+
+class WebhooksController < Telegram::Bot::UpdatesController
+  include StartContext
```
Comment (DmitryAvramec) [inline 3529/2/telegram.rb]: Why do I need to include all the staff here?

File: 3529/2/telegram.rb [hunk 22]
```diff
@@ -0,0 +1,30 @@
+require 'telegram/bot'
+require 'yaml'
+require 'httparty'
+require 'open-uri'
+require_relative 'start_context'
```
Comment (DmitryAvramec) [inline 3529/2/telegram.rb]: You can require all files with 1 line

File: 3529/2/telegram.rb [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+  end
+end
+
+TOKEN = '919190207:AAFfJhW2frNEWYeaSAvxhgwC6I233JVnVBg'.freeze
+bot = Telegram::Bot::Client.new(TOKEN)
+Telegram.bots_config = {
+  default: TOKEN
+}
+
+require 'logger'
```
Comment (DmitryAvramec) [inline 3529/2/telegram.rb]: to the top

File: 3529/2/number_checker.rb [hunk 24]
```diff
@@ -0,0 +1,44 @@
+require 'yaml'
+require 'redis'
+require 'redis-activesupport'
+require 'redis-rails'
+require 'redis-namespace'
+require 'redis-rack-cache'
```
Comment (DmitryAvramec) [inline 3529/2/number_checker.rb]: Why this?

File: 3529/2/checkin_context.rb (hunk 2 above)
Comment (gannagoodkevich) [reply 3529/2/checkin_context.rb]: fixing hound:(

File: 3529/2/telegram.rb (hunk 22 above)
Comment (gannagoodkevich) [reply 3529/2/telegram.rb]: it's about require_all? so i need to put all files i use in a one dir, am i right?

File: 3529/2/selfie_handler.rb (hunk 16 above)
Comment (gannagoodkevich) [reply 3529/2/selfie_handler.rb]: as far as i understend task, we need to do it, 'cause with this prog we find names of directory (we need to know session key and time)
or i'm confused

File: 3529/2/telegram.rb (hunk 22 above)
Comment (DmitryAvramec) [reply 3529/2/telegram.rb]: 1 way: 
https://coderwall.com/p/sfokrw/require-all-files-within-a-directory-in-ruby
Find the best solution for your project

File: 3529/2/number_checker.rb (hunk 13 above)
Comment (gannagoodkevich) [reply 3529/2/number_checker.rb]: if users number exist in database of camp i need to check if there is signed in user with this number

File: 3529/2/number_checker.rb (hunk 12 above)
Comment (gannagoodkevich) [reply 3529/2/number_checker.rb]: i need it for udating yaml file in update_file method here

File: 3529/2/telegram.rb (hunk 21 above)
Comment (gannagoodkevich) [reply 3529/2/telegram.rb]: for creating message context?)

File: 3529/2/telegram.rb (hunk 22 above)
Comment (gannagoodkevich) [reply 3529/2/telegram.rb]: thanks)))

File: 3529/2/telegram.rb (hunk 21 above)
Comment (DmitryAvramec) [reply 3529/2/telegram.rb]: I'm talking about includes all these modules when almost all of them include each other
Mb good idea is creating come `SomeStaffHelper` that contains all of one kind modules and inlcude it everywhere.

Comment (gannagoodkevich) [reply 3529/2/telegram.rb]: okey, sounds better, then i had)

File: 3529/2/checkin_context.rb (hunk 4 above)
Comment (gannagoodkevich) [reply 3529/2/checkin_context.rb]: ??? 
that's my wau to hanlde this stuff (about how to communicate with user):(
I really don't know how can i make this better

File: 3529/2/image_loader.rb (hunk 7 above)
Comment (gannagoodkevich) [reply 3529/2/image_loader.rb]: Chould i make them variables?

**Comment by gannagoodkevich** [comment]
Made a commit of resolved requests

File: 3529/2/image_loader.rb (hunk 7 above)
Comment (gannagoodkevich) [reply 3529/2/image_loader.rb]: i didn't get it
:(

File: 3529/2/telegram.rb [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+  include ContextHandler
+  include Telegram::Bot::UpdatesController::MessageContext
+
+  def initialize(_one, _two)
+    super
+    Telegram::Bot::UpdatesController.session_store = :redis_store, { expires_in: 100_000_000 }
+  end
+end
+
+file_token = YAML.safe_load(File.read('Data/token.yaml'))
```
Comment (DmitryAvramec) [inline 3529/2/telegram.rb]: can we move all this logic bellow to the class and then call it from 1 file like `MySuppaSweetBot.start/run` ?

File: 3529/2/public/313528695:313528695/checkins/07:24:2019:05/gannagood.txt [hunk 26]
```diff
@@ -0,0 +1 @@
+1234567
```
Comment (DmitryAvramec) [inline 3529/2/public/313528695:313528695/checkins/07:24:2019:05/gannagood.txt]: why I can see this? `.gitignore` for public folder
And I think you have a bad structure of folders `rubizza_number`/`checkins/checkout`/`timestamp`/
That's what we need, cos it's rly hard to look at time spent for some person without your matching table

File: 3529/2/Lib/checkout_context.rb [hunk 27]
```diff
@@ -0,0 +1,19 @@
+require 'fileutils'
+require './Lib/Utiles/path_loader.rb'
```
Comment (DmitryAvramec) [inline 3529/2/Lib/checkout_context.rb]: Don't think you need this includes
But if so - include all files in this folder with 1 line

File: 3529/2/Lib/context_handler.rb [hunk 28]
```diff
@@ -0,0 +1,18 @@
+require './Lib/Utiles/selfie_handler.rb'
+require './Lib/checkin_context.rb'
+require './Lib/checkout_context.rb'
+
+module ContextHandler
+  include SelfieHandler
+  include StartContext
```
Comment (DmitryAvramec) [inline 3529/2/Lib/context_handler.rb]: why we need this includes?

File: 3529/2/Lib/Utiles/image_loader.rb [hunk 29]
```diff
@@ -0,0 +1,7 @@
+module ImageLoader
+  def load_image
+    file_id =  payload['photo'].first['file_id']
+    response = HTTParty.get("https://api.telegram.org/bot#{TOKEN}/getFile?file_id=#{file_id}")
+    "https://api.telegram.org/file/bot#{TOKEN}/#{response['result']['file_path']}"
```
Comment (DmitryAvramec) [inline 3529/2/Lib/Utiles/image_loader.rb]: All long strings to constant please

File: 3529/2/Lib/Utiles/number_checker.rb [hunk 30]
```diff
@@ -0,0 +1,40 @@
...
+              "I'm sorry, but someone already have signed in with #{number} number"
+            else
+              new_part
+            end
+    participant['telegram_id'] = payload['from']['id'] if @resp.eql?(new_part)
+    @resp
+  end
+
+  def update_file
+    data_base = Redis::Namespace.new('telegram-bot-app', redis: data_base)
```
Comment (DmitryAvramec) [inline 3529/2/Lib/Utiles/number_checker.rb]: Why you start new instance of redis here?

File: 3529/2/Lib/Utiles/path_loader.rb [hunk 31]
```diff
@@ -0,0 +1,15 @@
+require 'fileutils'
+
+class PathLoader
+  def initialize(user_id, chat_id)
+    @session_id = "#{user_id}:#{chat_id}"
```
Comment (DmitryAvramec) [inline 3529/2/Lib/Utiles/path_loader.rb]: Mb better to use some static number? like rubizza numbers? Fell free to just resolve this 1

File: 3529/2/Lib/Utiles/path_loader.rb [hunk 32]
```diff
@@ -0,0 +1,15 @@
...
+    @date = Time.now.strftime('%m:%d:%Y:%H')
+  end
+
+  def create_directory(in_or_out, file_name)
+    path = "public/#{@session_id}/check#{in_or_out}s/#{@date}/"
```
Comment (DmitryAvramec) [inline 3529/2/Lib/Utiles/path_loader.rb]: https://ruby-doc.org/stdlib-2.4.1/libdoc/fileutils/rdoc/FileUtils.html#method-c-mkdir_p
Read please documentation or look into source code. You don't need this

File: 3529/2/Lib/checkout_context.rb [hunk 33]
```diff
@@ -0,0 +1,19 @@
...
+  include ImageFileWriter
+  include CoordinatesHandler
+
+  def checkout!(message = nil, *)
+    @path_check = PathLoader.new(payload['from']['id'], payload['chat']['id'])
+    checko(StatementFactory.build(message))
+  end
+
+  def checko(message)
+    public_send("reply_to_#{message.class}".downcase.to_sym, @path_check, 'out')
```
Comment (DmitryAvramec) [inline 3529/2/Lib/checkout_context.rb]: OOOOOO!!!!!!!
Explain it to me, please))))

Comment (gannagoodkevich) [reply 3529/2/Lib/checkout_context.rb]: I found this article: http://designisrefactoring.com/2015/03/09/refactoring-away-a-conditional/
To make long story short that will choose dynamic the right method for each case, depending on wich class was created. i can call method by its name with the help of public_send method.

File: 3529/2/Lib/context_handler.rb (hunk 28 above)
Comment (gannagoodkevich) [reply 3529/2/Lib/context_handler.rb]: You mean that i should put all methods from modules just in this file?

**Comment by gannagoodkevich** [comment]
add commit with solved requests

File: 3529/2/Lib/Utiles/image_loader.rb (hunk 29 above)
Comment (gannagoodkevich) [reply 3529/2/Lib/Utiles/image_loader.rb]: I can't find a way how to do it in such situation. 'cause i need to pass hound test, and it requires freeze constants, but this constants should be changed depending on images... :(
So i don't change it in last commit

**Comment by gannagoodkevich** [comment]
add commit with solved requested changes

File: 3529/2/Lib/Utiles/image_loader.rb (hunk 29 above)
Comment (DmitryAvramec) [reply 3529/2/Lib/Utiles/image_loader.rb]: `"https://api.telegram.org/bot#{TOKEN}/getFile?file_id=` this part will never be changed

File: 3529/2/Lib/Utiles/utiles.rb [hunk 34]
```diff
@@ -0,0 +1,25 @@
...
+  def self.all
+    [Coordinates, Selfie]
+  end
+end
+
+Statement = Struct.new(:message)
+
+class Selfie < Statement
+  def self.match?(_message)
+    true
```
Comment (DmitryAvramec) [inline 3529/2/Lib/Utiles/utiles.rb]: Why `true` here?

File: 3529/2/Lib/Utiles/utiles.rb [hunk 35]
```diff
@@ -0,0 +1,25 @@
...
+  end
+end
+
+class Coordinates < Statement
+  def self.match?(message)
+    message
```
Comment (DmitryAvramec) [inline 3529/2/Lib/Utiles/utiles.rb]: And `message` here

File: 3529/2/Lib/Utiles/utiles.rb [hunk 36]
```diff
@@ -0,0 +1,25 @@
...
+  end
+end
+
+class StatementFactory
+  def self.build(message)
+    Statements.all.detect { |massage| massage.match?(message) }.new(message)
```
Comment (DmitryAvramec) [inline 3529/2/Lib/Utiles/utiles.rb]: `massage`? )

Comment (DmitryAvramec) [reply 3529/2/Lib/Utiles/utiles.rb]: Btw why you need this detect?

File: 3529/2/Lib/context_handler.rb [hunk 37]
```diff
@@ -0,0 +1,59 @@
...
+    public_send("reply_to_#{message.class}".downcase.to_sym, @path_check, 'out')
+    save_context :checkout!
+  end
+
+  def checkin!(message = nil, *)
+    @path_check = PathLoader.new(payload['from']['id'])
+    checki(StatementFactory.build(message))
+  end
+
+  def checki(message)
```
Comment (DmitryAvramec) [inline 3529/2/Lib/context_handler.rb]: Rename these 2 methods please

**Pull request closed without merging**